src/common/summarize-profile.js

```diff
@@ -110,17 +97,43 @@ export function summarizeProfile(profile: Profile) {
  */
 function functionNameCategorizer() {
   const cache = new Map();
+
+  // Compile set of categories into a single regular expression that produces
+  // the pattern (whether it's a prefix, stem, substring, or exact match).
+
+  const patternToCategory = new Map();
+
+  const matchPatterns = {
+    prefix: [],
+    stem: [],
+    substring: [],
+    exact: []
+  };
+
+  for (const [matchType, pattern, category] of categories) {
+    patternToCategory.set(pattern, category);
+    let pat = pattern.replace(/\W/g, '\\$&');
+    if (matchType == 'stem')
+      pat += "\\b(?=\\()?";
+    matchPatterns[matchType].push(pat);
+  }
+
+  const regex = new RegExp(  "^(?:"    + matchPatterns['prefix'].join("|") + ")"
+                           + "|(?:"    + matchPatterns['substring'].join("|") + ")"
+                           + "|\\b(?:" + matchPatterns['stem'].join("|") + ")"
+                           + "|^(?:"   + matchPatterns['exact'].join("|") + ")$");
```

Comment:
looking at our patterns, I feel like we could replace all stem and substring matches by a prefix match. What do you think @mstange ? (not for this patch though)

I feel like substring matches is slower than the others (because there's no anchor) and if we can eliminate it we should see a speed boost. I don't know if this would be noticeable :D